File: lib/tmdb_api/model/movie_details.dart
import 'dart:core';

import 'package:tmdb/tmdb_api/credits_response.dart';

import 'genre.dart';
import 'language.dart';
import 'movie.dart';
import 'movie_status.dart';
import 'production_company.dart';
import 'production_country.dart';

class MovieDetails extends Movie {
  int budget;
  CreditsResponse credits;
  List<Genre> genres;
  String homepage;
  String imdbId;
  Movie movie;
  List<ProductionCompany> productionCompanies;
  List<ProductionCountry> productionCountries;
  int revenue;
  int runtime;
  List<SpokenLanguage> spokenLanguages;
  MovieStatus status;
  String tagline;

  MovieDetails({
    this.budget,
    this.genres,
    this.homepage,
    this.imdbId,
    this.movie,
    this.productionCompanies,
    this.productionCountries,
    this.revenue,
    this.runtime,
    this.spokenLanguages,
    this.status = MovieStatus.released,
    this.tagline,
    this.credits,
  })  :super(
          backdropPath: movie.backdropPath,
          genreIds: movie.genreIds,
          media: movie.media,
          originalLanguage: movie.originalLanguage,
          originalTitle: movie.originalTitle,
          overview: movie.overview,
          posterPath: movie.posterPath,
          releaseDate: movie.releaseDate,
          title: movie.title,
          video: movie.video,
          voteAverage: movie.voteAverage,
          voteCount: movie.voteCount,
        );

  MovieDetails.of(Movie movie) : this(movie: movie);

  /// Creates a [MovieDetails] from a JSON object.
  factory MovieDetails.fromJson(Map<String, dynamic> json) {
    if (json == null) return null;

    var list = json['genres'] as List;
    List<Genre> genres = list.map((i) => Genre.fromJson(i)).toList();

    list = json['production_companies'] as List;
    List<ProductionCompany> productionCompanies =
        list?.map((i) => ProductionCompany.fromJson(i))?.toList();

    list = json['production_countries'] as List;
    List<ProductionCountry> productionCountries =
        list?.map((i) => ProductionCountry.fromJson(i))?.toList();

    list = json['spoken_languages'] as List;
    List<SpokenLanguage> spokenLanguages =
        list.map((i) => SpokenLanguage.fromJson(i)).toList();

    return MovieDetails(
      budget: json['budget'],
      credits: CreditsResponse.fromJson(json['credits']),
      genres: genres,
      homepage: json['homepage'],
      imdbId: json['imdb_id'],
      movie: Movie.fromJson(json),
      productionCompanies: productionCompanies,
      productionCountries: productionCountries,
      revenue: json['revenue'],
      runtime: json['runtime'],
      spokenLanguages: spokenLanguages,
      status: MovieStatus.valueOf(json['status']),
      tagline: json['tagline'],
    );
  }
}

File: android/model/src/main/java/com/tikal/tmdb/model/MovieDetails.kt
package com.tikal.tmdb.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.Calendar

/**
 * Movie details POJO.
 */
@Serializable
class MovieDetails : Movie {
    @SerialName("budget")
    val budget: Int

    @SerialName("homepage")
    val homepage: String?

    @SerialName("imdb_id")
    val imdbId: String?

    @SerialName("production_companies")
    val productionCompanies: List<ProductionCompany>?

    @SerialName("production_countries")
    val productionCountries: List<ProductionCountry>?

    @SerialName("revenue")
    val revenue: Int?

    @SerialName("runtime")
    val runtime: Int?

    @SerialName("spoken_languages")
    val spokenLanguages: List<SpokenLanguage>

    @SerialName("status")
    val status: String

    @SerialName("tagline")
    val tagline: String?

    constructor(
        id: Long,
        adult: Boolean,
        backdropPath: String?,
        budget: Int = 0,
        genres: List<Genre> = emptyList(),
        homepage: String? = null,
        imdbId: String? = null,
        originalLanguage: String,
        originalTitle: String,
        overview: String?,
        popularity: Float,
        posterPath: String?,
        productionCompanies: List<ProductionCompany>? = null,
        productionCountries: List<ProductionCountry>? = null,
        releaseDate: Calendar?,
        revenue: Int? = null,
        runtime: Int? = null,
        spokenLanguages: List<SpokenLanguage> = emptyList(),
        status: String = "",
        tagline: String? = null,
        title: String,
        video: Boolean,
        voteAverage: Float,
        voteCount: Int
    ) : super(
        id = id,
        adult = adult,
        backdropPath = backdropPath,
        genreIds = genres.map { it.id }.toLongArray(),
        originalLanguage = originalLanguage,
        originalTitle = originalTitle,
        overview = overview,
        popularity = popularity,
        posterPath = posterPath,
        releaseDate = releaseDate,
        title = title,
        video = video,
        voteAverage = voteAverage,
        voteCount = voteCount
    ) {
        this.budget = budget
        this.homepage = homepage
        this.imdbId = imdbId
        this.productionCompanies = productionCompanies
        this.productionCountries = productionCountries
        this.revenue = revenue
        this.runtime = runtime
        this.spokenLanguages = spokenLanguages
        this.status = status
        this.tagline = tagline
    }

    companion object {
        fun of(movie: Movie): MovieDetails {
            return MovieDetails(
                id = movie.id,
                adult = movie.adult,
                backdropPath = movie.backdropPath,
                genres = movie.genreIds.map { genreId -> Genre(genreId, genreId.toString()) },
                originalLanguage = movie.originalLanguage,
                originalTitle = movie.originalTitle,
                overview = movie.overview,
                popularity = movie.popularity,
                posterPath = movie.posterPath,
                releaseDate = movie.releaseDate,
                title = movie.title,
                video = movie.video,
                voteAverage = movie.voteAverage,
                voteCount = movie.voteCount
            )
        }
    }
}